BuildMasterAutomation/Functions/ConvertFrom-BMOtterScriptExpression.ps1

```diff
@@ -0,0 +1,189 @@
+            )
+
+            $maybeInt = 0
+            if (([Int64]::TryParse($Item, [ref] $maybeInt)))
+            {
+                return $maybeInt
+            }
+
+            return $Item
+        }
+    }
+
+    process {
+        $isMap = $Value.StartsWith('%(') -and $Value.EndsWith(')')
+        $isVector = $Value.StartsWith('@(') -and $Value.EndsWith(')')
+        $originalValue = $Value
+
+        if (-not $isMap -and -not $isVector)
+        {
+            return ConvertTo-Int -Item $Value.Trim()
+        }
+
+        if ($isMap)
+        {
+            $Value = $Value -replace '^%\('
+        }
+        else
+        {
+            $Value = $Value -replace '^@\('
+        }
+
+        $Value = $Value -replace '\)$'
+
+        $closesNeeded = 0
+        $failingSyntax = $false
+        # Splitting up array or map by comma and collecting into array.
+        $parsedItems = &{
+            $start = 0
+            $end = 0
+            foreach ($i in 0..($Value.Length - 1))
+            {
+                $char = $Value[$i]
+
+                if ($char -eq '(' -and ($Value[$i - 1] -eq '@' -or $Value[$i - 1] -eq '%'))
+                {
+                    $closesNeeded++
+                    continue
+                }
+
+                if ($char -eq ')' -and $closesNeeded)
+                {
+                    $closesNeeded--
+                    if ($i -ne ($Value.Length - 1))
+                    {
+                        continue
+                    }
+                }
+
+                if ($closesNeeded -or ($char -ne ',' -and $i -ne ($Value.Length - 1)))
+                {
+                    continue
+                }
+
+                $end = $i
+                $lengthOfSubstring = $end - $start
+
+                if ($i -eq ($Value.Length - 1))
+                {
+                    $lengthOfSubstring++
+                }
+
+                $entry = $Value.Substring($start, $lengthOfSubstring).Trim()
+                $start = $i + 1
+
+                if ($isVector)
+                {
+                    $converted = ConvertFrom-BMOtterScriptExpression -Value $entry
+
+                    if ($converted -is [array])
+                    {
+                        # PowerShell Array flattening is the bane of my existence...
+                        @{ 'Item' = $converted } | Write-Output
+                        continue
+                    }
+                    $converted | Write-Output
+                    continue
+                }
+
+                $firstColonIndex = $entry.IndexOf(':')
+                if ($firstColonIndex -lt 0)
+                {
+                    $failingSyntax = $true
+                    break
+                }
+
+                @{
+                    'Key' = ConvertTo-Int -Item $entry.Substring(0, $firstColonIndex).Trim()
+                    'Value' = ConvertFrom-BMOtterScriptExpression -Value $entry.Substring($firstColonIndex + 1).Trim()
+                } | Write-Output

```

Comment:
This loop is just for parsing the individual items in a vector or map expression. It shouldn't be doing any recursive calls here, instead do that below once all the individual items are resolved.